### PR TITLE
Improve Skipgram/Paragraph Vectors  performance 

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
Work on improving skipgram performance. There was an overall regression in performance from earlier versions after migrating to the c++ op.

Also introduces a new serialization framework to reduce GC pressure when saving word2vec like models.

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [X ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ X] Created tests for any significant new code additions.
- [ X] Relevant tests for your changes are passing.

### CODE REVIEW

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/factory/Nd4j.java [hunk 1]
```diff
@@ -2974,6 +2974,19 @@ public static INDArray rand(int[] shape, @NonNull org.nd4j.linalg.api.rng.Random
...
+
+    /**
+     * Create a random ndarray with the given shape using the given RandomGenerator
+     *
+     * @param shape the shape of the array
+     * @param rng     the random generator to use
+     * @return the random ndarray with the specified shape
+     */
+    public static INDArray rand(@NonNull org.nd4j.linalg.api.rng.Random rng, DataType dataType,@NonNull long... shape) {
+        INDArray ret = createUninitialized(shape, Nd4j.order()).castTo(dataType); //INSTANCE.rand(shape, rng);
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/factory/Nd4j.java]: Please don't leave any left-over commented out code here.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/factory/Nd4j.java [hunk 2]
```diff
@@ -5615,7 +5657,22 @@ public static void writeAsNumpy(INDArray arr, OutputStream writeTo) throws IOExc
             }
 
             bufferedOutputStream.flush();
+            return written;
         }
+
+    }
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/factory/Nd4j.java:5620]: That looks quite repetitive, no way to rewrite this such that the `closeFlush` flag doesn't produce duplicate code?

File: platform-tests/src/test/java/org/deeplearning4j/models/paragraphvectors/ParagraphVectorsTest.java [hunk 3]
```diff
@@ -164,193 +165,218 @@ public void testParagraphVectorsVocabBuilding1() throws Exception {
...
+            ParagraphVectors paragraphVectors = binary ?
+                    WordVectorSerializer.readParagraphVectorsBinary(fullFile) :
+                    WordVectorSerializer.readParagraphVectors(fullFile);
+
+            //TODO: figure out why some json properties are not saved properly
+          /*  assertEquals(vec.getConfiguration(),paragraphVectors.getConfiguration());
+            assertEquals(vec.vocab(),paragraphVectors.vocab());
+            assertEquals(vec.lookupTable(),paragraphVectors.lookupTable());
+            assertEquals(vec.getLayerSize(),paragraphVectors.getLayerSize());
+     */       //assertEquals(vec,paragraphVectors);
```
Comment (treo) [inline platform-tests/src/test/java/org/deeplearning4j/models/paragraphvectors/ParagraphVectorsTest.java]: Was this TODO handled?

**Review by treo — CHANGES_REQUESTED** [review]

File: platform-tests/src/test/java/org/deeplearning4j/models/paragraphvectors/ParagraphVectorsTest.java (hunk 3 above)
Comment (agibsonccc) [reply platform-tests/src/test/java/org/deeplearning4j/models/paragraphvectors/ParagraphVectorsTest.java]: Yes it was. Did clean up.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/factory/Nd4j.java (hunk 1 above)
Comment (agibsonccc) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/factory/Nd4j.java]: Both instances here were old code. I cleaned that up and replaced all duplicate usages of redundant in the tests.

Comment (treo) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/factory/Nd4j.java]: Please also remove the `//INSTANCE.rand(shape, rng);` in this line :)

**Review by treo — CHANGES_REQUESTED** [review]

**Review by treo — APPROVED** [review]

**Pull request merged**